File: java/src/main/java/br/com/atlas/averbacao/api/ProblemException.java
package br.com.atlas.averbacao.api;

public class ProblemException extends RuntimeException {
    private final int status;
    private final String title;

    public ProblemException(int status, String title, String detail) {
        super(detail);
        this.status = status;
        this.title = title;
    }

    public int status() { return status; }
    public String title() { return title; }
}

File: dotnet/Atlas.Averbacao/Api/ProblemException.cs
namespace Atlas.Averbacao.Api;

public class ProblemException : Exception
{
    public int Status { get; }
    public string Title { get; }

    public ProblemException(int status, string title, string detail) : base(detail)
    {
        Status = status;
        Title = title;
    }
}

public class NotImplementedProblem : ProblemException
{
    public NotImplementedProblem(string operation) : base(501, "Not Implemented", operation) { }
}

public class UnauthorizedProblem : ProblemException
{
    public UnauthorizedProblem(string detail) : base(401, "Unauthorized", detail) { }
}
